### PR TITLE
Use fmt::Display to format error messages in logs

### PR DESCRIPTION
**Pull request by hawkw**
This PR changes the proxy to log error messages using `fmt::Display` whenever possible, which should lead to much more readable and meaningful error messages

This is part of the work I started last week on issue #442. While I haven't finished everything for that issue (all errors still are mapped to HTTP 500 error codes), I wanted to go ahead and open a PR for the more readable error messages. This is partially because I found myself merging these changes into other branches to aid in debugging, and because I figured we may as well have the nicer logging on master.

### CODE REVIEW

File: proxy/src/map_err.rs [hunk 1]
```diff
@@ -1,50 +1,81 @@
...
 
 /// Catches errors from the inner future and maps them to 500 responses.
-pub struct ResponseFuture<T, E> {
+pub struct ResponseFuture<T, E, F> {
     inner: T,
+    f: Arc<F>,
     _p: PhantomData<E>,
 }
 
+// pub fn get_cause<E>(whatever: &E) -> Option<&Error> {
```
Comment (seanmonstar) [inline proxy/src/map_err.rs]: Can be removed.

File: proxy/src/map_err.rs [hunk 2]
```diff
@@ -1,50 +1,81 @@
...
 
-impl<T, B, E> Service for MapErr<T, E>
+impl<T, E> MapErr<T, E, fn(E) -> http::StatusCode>
+where
+    T: Service<Error = E>,
+    E: Debug,
+{
+    /// Crete a new `MapErr` with the default behaviour
+    /// (mapping all errors to HTTP 500).
+    #[allow(dead_code)]
```
Comment (seanmonstar) [inline proxy/src/map_err.rs]: Whats the `dead_code` lint triggering on?

File: proxy/src/lib.rs [hunk 3]
```diff
@@ -317,8 +318,31 @@ where
...
         let router = router.clone();
 
-        // Map errors to 500 responses
-        MapErr::new(router)
+        // Map errors to appropriate response error codes.
+        MapErr::new(router, |e| {
+
+            match e {
+                RouteError::Route(r) => {
+                    error!("route error: {:?}", r);
```
Comment (seanmonstar) [inline proxy/src/lib.rs]: Did this mean to use `Display`?

File: proxy/src/lib.rs [hunk 4]
```diff
@@ -317,8 +318,31 @@ where
...
+                    http::StatusCode::INTERNAL_SERVER_ERROR
+                }
+                RouteError::Inner(i) => {
+                    // TODO: type system hates this...
+                    // if let Some(cause) = i.cause() {
```
Comment (seanmonstar) [inline proxy/src/lib.rs]: I don't get what the type system hates...

**Comment by hawkw** [comment]
@seanmonstar sorry for the mess, I was working on mapping more errors to specific error codes & commented out the work in progress code rather than deleting it. c25133777fc2358d797fc98758d746452f5b4562 should clean it up a bit!

File: proxy/src/lib.rs [hunk 5]
```diff
@@ -317,8 +318,23 @@ where
...
+                    error!("route error: {}", r);
+                    http::StatusCode::INTERNAL_SERVER_ERROR
+                }
+                RouteError::Inner(i) => {
+                    error!("{}", i);
+                    http::StatusCode::INTERNAL_SERVER_ERROR
+                }
+                RouteError::NotRecognized => {
+                    error!("route not recognized");
+                    http::StatusCode::NOT_FOUND
```
Comment (seanmonstar) [inline proxy/src/lib.rs]: What instances does this `NotRecognized` error occur? Is it true that it is **always** the user's fault?

Comment (hawkw) [reply proxy/src/lib.rs]: Hmm, I'll have to take a closer look at that.

Comment (hawkw) [reply proxy/src/lib.rs]: I believe that `NotRecognized` is returned in the following cases:

+ By the outbound side of the proxy, if the request had no authority, no host header, and no SO_ORIGINAL_DST (or, we were on an OS that doesn't support SO_ORIGINAL_DST):
https://github.com/runconduit/conduit/blob/82bbcbd13769227aac7a1354b4106d6818f671e7/proxy/src/outbound.rs#L83-L85

+ By the inbound side of the proxy if the request had no SO_ORIGINAL_DST and the inbound was not configured with a default address.

I'm not sure if 404 is the correct error code to return in these cases or not --- what do you think? Are these bad requests?

Comment (hawkw) [reply proxy/src/lib.rs]: @seanmonstar as of 06a7b8f3be590347c4bda8514b6db208b3062ee9, everything gets 500 errors once again. Would still like to return more accurate error codes in a future PR, but this is the most correct behavior for now, and I'd like to get the improved logging merged.

**Pull request merged**